ext/PathfinderTuringExt.jl

```diff
@@ -101,6 +103,58 @@ function varnames_to_ranges(metadata::DynamicPPL.Metadata)
     return Dict(zip(metadata.vns, ranges))
 end
 
+"""
+    transform_to_constrained(
+        p::AbstractArray, vi::DynamicPPL.VarInfo, model::DynamicPPL.Model
+    )
+
+Transform a vector of parameters `p` from unconstrained to constrained space.
+"""
+function transform_to_constrained(
+    p::AbstractArray, vi::DynamicPPL.VarInfo, model::DynamicPPL.Model
+)
+    p = copy(p)
+    @assert DynamicPPL.istrans(vi)
+    vi = DynamicPPL.unflatten(vi, p)
+    p .= DynamicPPL.invlink!!(vi, model)[:]
+    # Restore the linking, since we mutated vi.
+    DynamicPPL.link!!(vi, model)
+    return p
+end
+
+"""
+    set_up_model_optimisation(model::DynamicPPL.Model, init)
+
+Create the necessary pieces for running optimisation on `model`.
+
+Returns
+* An `Optimization.OptimizationFunction` that evaluates the log density of the model and its
+gradient in the unconstrained space.
+* The initial value `init` transformed to unconstrained space.
+* A function `transform_result` that transforms the results back to constrained space. It
+takes a single vector argument.
+"""
+function set_up_model_optimisation(model::DynamicPPL.Model, init)
+    # The inner context deterimines whether we are solving MAP or MLE.
+    inner_context = DynamicPPL.DefaultContext()
+    ctx = Turing.Optimisation.OptimizationContext(inner_context)
+    log_density = Turing.Optimisation.OptimLogDensity(model, ctx)
+    # Initialise the varinfo with the initial value and then transform to unconstrained
+    # space.
+    Accessors.@set log_density.varinfo = DynamicPPL.unflatten(log_density.varinfo, init)
+    transformed_varinfo = DynamicPPL.link(log_density.varinfo, log_density.model)
+    log_density = Accessors.@set log_density.varinfo = transformed_varinfo
+    init = log_density.varinfo[:]
+    # Create a function that applies the appropriate inverse transformation to results, to
+    # bring them back to constrained space.
+    transform_result(p) = transform_to_constrained(p, log_density.varinfo, model)
+    f = Optimization.OptimizationFunction(
+        (x, _) -> log_density(x),;
+        grad = (G,x,p) -> log_density(nothing, G, x),
+    )
+    return f, init, transform_result
```

Comment:
`f` here seems to neglect the logdetjac term of the bijector. e.g.

```julia
julia> using Pathfinder, Turing

julia> PathfinderTuringExt = Base.get_extension(Pathfinder, :PathfinderTuringExt);

julia> dist = Dirichlet(ones(5));

julia> xs = [rand(dist), rand(dist)];

julia> logpdf.(Ref(dist), xs)  # constant density
2-element Vector{Float64}:
 3.1780538303479458
 3.1780538303479458

julia> dist_trans = Bijectors.transformed(dist);

julia> ys = Bijectors.bijector(dist).(xs)
2-element Vector{Vector{Float64}}:
 [-0.24602072829084776, 0.1511670612099265, -1.7070748562188633, -2.5572029034842583]
 [-3.1526698251090055, 1.2795646828995244, 0.17138006669735018, -0.9312267844619441]

julia> logpdf.(Ref(dist_trans), ys)  # non-constant due to logdet jacobian
2-element Vector{Float64}:
 -6.972313530283863
 -7.899132706003789

julia> @model function demo()
    x ~ dist
end;

julia> f, init, transform_result = PathfinderTuringExt.set_up_model_optimisation(demo(), xs[1]);

julia> f.(ys, nothing)  # constant density, logdet jacobian is missing
2-element Vector{Float64}:
 -3.1780538303479458
 -3.1780538303479458
```

Reply:
Isn't this what one would want for optimisation purposes though? For instance, in the above case, say we want to find the maximum of `dist`. The `f` that returns a constant value would correctly yield zero gradients, saying "doesn't matter where you are in the parameter space, it's all flat, nothing to optimise", whereas the version that includes the Jacobian term would start to guide the optimisation towards some particular point (the maximum of the log det Jacobian term).

Reply:
For purely optimization, yes, but we are not just or even necessarily trying to find the MAP. We use the optimization trajectory to propose a series of variational distribution approximations we then estimate the ELBO for. These are multivariate normal. So we definitely want the target distribution to represent the posterior in some space, and for the draws from the variational distribution to obey any constraints on the posterior, we need to compute the ELBO relative to the unconstrained posterior, which requires the Jacobian term. The ELBO itself is invariant under simultaneous reparameterization of both the target and variational distributions, so we effectively are then approximating the posterior with a multivariate normal pushed through the constraining transform.

Reply:
I'll need to read what you wrote again next week to absorb it fully, but I wonder if what you want is just dropping the `OptimizationContext` and replacing it with `DefaultContext`.

Note that this ignoring of the Jacobian of the transformation is what `optim_function` did in Turing v0.32 too, there's probably been a bug here all along.

Reply:
> I wonder if what you want is just dropping the `OptimizationContext` and replacing it with `DefaultContext`.

That seems to be a problem here: https://github.com/TuringLang/Turing.jl/blob/f13582124a38f7af85b4e3b77b43f47a53dfc240/src/optimisation/Optimisation.jl#L107. `OptimLogDensity` is a `LogDensityFunction` with the context fixed to an `OptimizationContext`. We can easily create the `LogDensityFunction`, but it's only callable as an `OptimLogDensity`, e.g. https://github.com/TuringLang/Turing.jl/blob/f13582124a38f7af85b4e3b77b43f47a53dfc240/src/optimisation/Optimisation.jl#L130-L138

Reply:
Considering that the use case here is substantially different from MAP/MLE, I wouldn't feel bad just replicating much of the functionality of `OptimLogDensity` in a new type/function in Pathfinder that uses plain `DefaultContext`. Might also simplify the gradient stuff a bit in the process, `OptimLogDensity` currently caters to the needs of Optim.jl, but with Optimisation.jl I think we can leave taking the gradient to Optimisation.jl's AD system.

I can take a look at it, but I won't have time today. Feel free to do it too if you have time.

Also, having a test that catches the issue of ignoring the Jacobian would be great.